kernel/src/schema.rs

```diff
@@ -188,32 +186,27 @@ impl StructField {
             .collect()
     }
 
-    pub fn make_physical(&self, mapping_mode: ColumnMappingMode) -> DeltaResult<Self> {
-        use ColumnMappingMode::*;
-        match mapping_mode {
-            Id => return Err(Error::generic("Column ID mapping mode not supported")),
-            None => return Ok(self.clone()),
-            Name => {} // fall out
-        }
-
-        struct ApplyNameMapping;
-        impl<'a> SchemaTransform<'a> for ApplyNameMapping {
+    /// Applies physical name mappings to this field
+    ///
+    /// NOTE: Caller affirms that the schema was alreasdy validated by
+    /// [`crate::table_features::validate_schema_column_mapping`], to ensure that
+    /// annotations are always and only present when column mapping mode is enabled.
+    pub fn make_physical(&self) -> Self {
```

Comment:
(tell me if this is overkill)

In an ideal world I think we would encode this in the type system so that it's prohibited to ever have a `make_physical()` call without calling `validate_schema_column_mapping`. I was trying to think of relatively easy ways to encode this and came up with two ideas (one easy, one harder)

1. (easy) rename `Schema` before it is validated to `struct UnvalidatedSchema(Schema)` and require a validate method to consume the unvalidated and produce a regular `Schema`. Considering the approach of this PR is to do the validation up-front this seems to be the easiest?
2. (harder) We could leverage a couple zero-sized types to tag a `Schema` as either validated/unvalidated. Something like:
```rust
// markers
pub struct Validated;
pub struct Unvalidated;

pub struct Schema<ValidationState = Unvalidated> {
    // ...
    _validation_state: PhantomData<ValidationState>,
}

// then a function on `Schema<Unvalidated>` to create a `Schema<Validated>`
impl Schema<Unvalidated> {
    pub fn validate(self, mode: ColumnMappingMode) -> DeltaResult<Schema<Validated>> {
        // do column mapping validation and other validations in the future
    }
}
```

Reply:
Very interesting question. I actually have started wondering the same thing about physical vs. logical schema -- it's too easy to mix them up, as evidenced by the data skipping code that was trying to apply an expression full of logical column names to a physical parquet schema. And IIRC, delta-spark has hit more than one bug where somebody forgot which schema they were working with (and there are probably places where they double-apply the logical->physical mapping as well, tho that should be harmless). Probably a pair of newtypes would be the best way to make it explicit: `struct PhysicalSchema(StructType)` and `struct LogicalSchema(StructType)`. TBD whether they should impl `Deref` or `AsRef` -- for convenience they probably should, and anyway it would be a red flag for some class or method that cares about the difference to take a bare `StructType`.

For this specific case -- the unvalidated schema should be _very_ short-lived. If we changed `Metadata::schema` (**) to return a newtype `UnvalidatedSchema` with the `validate` method you suggest, then nobody else would have to know or care about it -- all bare schemas are known to be validated. We'd have to be careful on the write path any time we modify a schema in a way that could have implications for column mapping validation (create table, add column, enable column mapping, etc). But those operations would just need to produce an `UnvalidatedSchema` as their output?

(**) BTW, we should rename that method as `Metadata::parse_schema` to make clear that it's an expensive function to call -- not a getter.

Reply:
Update: After trying out the `UnvalidatedSchema` concept locally, I learned that the validation of protocol, schema, and table properties are all intertwined, so it's not enough to merely validate the schema. Most likely, we'll need to introduce a new `TableConfiguration` that encapsulates all three (and eventually system domain metadata as well), which cross-checks everything and with helpful utility methods to expose e.g. column mapping mode. 

IMO that's a bigger effort for a different PR.

Reply:
yep sounds good! I liked the `TableConfiguration` idea!!